[docs] Add runtime-audit-engine module architecture section (#20613)
deckhouse/deckhouse: docs/documentation/pages/architecture/security/RUNTIME-AUDIT-ENGINE.md

---
title: Runtime-audit-engine module
permalink: en/architecture/security/runtime-audit-engine.html
search: security audit, audit rules, falco, runtime-audit-engine
description: Architecture of the runtime-audit-engine module in Deckhouse Kubernetes Platform.
---

> The [`runtime-audit-engine`](/modules/runtime-audit-engine/) module implements [security event auditing](./runtime-audit.html) in Deckhouse Kubernetes Platform (DKP) based on the [Falco](https://falco.org/) threat detection system. The module collects Linux kernel events and Kubernetes API audit events (using the `k8saudit` plugin), enriches them with Kubernetes Pod metadata, and generates security events according to configured rules. Audit rules are defined using the [FalcoAuditRules](/modules/runtime-audit-engine/cr.html#falcoauditrules) custom resource.

Reviewer: ```suggestion
The [`runtime-audit-engine`](/modules/runtime-audit-engine/) module implements [runtime auditing](./runtime-audit.html) in Deckhouse Kubernetes Platform (DKP) based on the [Falco](https://falco.org/) threat detection system. The module collects Linux kernel events and Kubernetes API audit events (using the `k8saudit` plugin), enriches them with Kubernetes Pod metadata, and generates security events according to configured rules. Audit rules are defined using the [FalcoAuditRules](/modules/runtime-audit-engine/cr.html#falcoauditrules) custom resource.
```

- Pods may run multiple replicas. However, each pod is shown as a single replica in the diagram.
{% endalert %}

The Level 2 C4 architecture of the [`runtime-audit-engine`](/modules/runtime-audit-engine/) module and its interactions with other DKP components are shown in the following diagram:

<!--- Source: structurizr code from https://fox.flant.com/team/d8-system-design/doc/-/tree/main/architecture/diagrams/C4_EN --->
![Runtime-audit-engine module architecture](../../images/architecture/security/c4-l2-runtime-audit-engine.svg)

## Module components

The `runtime-audit-engine` module consists of the following components:

1. **Runtime-audit-engine** (DaemonSet): A component deployed on each cluster node. It collects audit events, evaluates rules, and exports triggered rules as Prometheus metrics. Data is received from the Linux kernel via an eBPF driver and from `containerd` via a Unix socket.

   The component includes the following containers:

>    - **falco**: Main container that collects security events from cluster nodes and containerized applications in DKP based on the [Falco](https://falco.org/) threat detection system. Falco intercepts Linux kernel system calls (syscalls) in real time, processes them based on rules, and outputs triggered audit rule results to stdout.

Author: ```suggestion
1. **Runtime-audit-engine** (DaemonSet): A component deployed on each cluster node. It collects audit events, evaluates rules, outputs triggered rules to stdout, and exports them as Prometheus metrics. Data is received from the Linux kernel via interception of system calls (syscalls) and from `containerd` via a Unix socket.

   The component includes the following containers:

   - **falco**: Main container that collects security events from cluster nodes and containerized applications in DKP based on the [Falco](https://falco.org/) threat detection system. 
```

   - **falcosidekick**: Sidecar container that receives events from the `falco` component and exports audit events as Prometheus metrics.
   - **rules-loader**: Sidecar container that performs the following operations:
      - watches [FalcoAuditRules](/modules/runtime-audit-engine/cr.html#falcoauditrules) custom resources and stores them in the shared `/etc/falco/rules.d/` Pod directory for processing by the `falco` component;

>       - validates the FalcoAuditRules custom resource.

Reviewer: ```suggestion
      - Watches [FalcoAuditRules](/modules/runtime-audit-engine/cr.html#falcoauditrules) custom resources and stores them in the shared `/etc/falco/rules.d/` Pod directory for processing by the `falco` component.
      - Validates the FalcoAuditRules custom resource.
```
